File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from collections import  deque
def bfs(x,y):
    global area
    queue=deque()
    queue.append((x,y))

    while queue :
        x,y=queue.popleft()

        for i in range (4):
            nx=x+dx[i]
            ny=y+dy[i]

            if 0<=nx<m and 0<=ny<n :
                if not visited[nx][ny] and graph[nx][ny] == 0:
                    visited[nx][ny]=True
                    queue.append((nx,ny))
                    area+=1
    area_cnt.append(area)


m,n,k=map(int,input().split())
graph=[[0]*n for _ in range(m)]

for i in range(k):
    x1,y1,x2,y2=map(int,input().split())

    for j in range (y1,y2):
        for k in range (x1,x2):
            graph[j][k] =1

dx=[1,-1,0,0]
dy=[0,0,1,-1]

visited=[[False]*n for _ in range(m)]

date=0
area=0
area_cnt=[]
for i in range (m):
    for j in range (n):
        if graph[i][j] == 0 and not visited[i][j]:
            visited[i][j]=True
            date+=1
            area=1
            bfs(i,j)

print(date)
area_cnt.sort()
for i in area_cnt:
    print(i,end=' ')

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.*;
import java.util.*;

public class Main {
	static int m;
	static int n;
	static int k;
	static int[][] graph;
	static boolean[][] visited;
 	static int[] dx = {1,-1,0,0};
	static int[] dy = {0,0,1,-1};
	
	static int cnt;
	static List<Integer> result = new ArrayList<>();
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();
		m = Integer.parseInt(st.nextToken());
		n =Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		
		graph = new int[m][n];
		visited = new boolean[m][n];
		cnt = 0;
		
		for (int i=0; i<k; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			int d = Integer.parseInt(st.nextToken());
			ice(a,b,c,d);
			
		}
		
		for (int i = 0; i<m; i++) {
			for (int j =0; j<n; j++) {
				if(graph[i][j] == 0 && !visited[i][j]) {
					cnt++;
					bfs(i, j);
				}
			}
		}
		System.out.println(cnt);
		result.sort(null);
		
		for(int i=0; i<result.size(); i++) {
			sb.append(result.get(i) + " ");
		}
		System.out.println(sb);
		
		
	}
	static void ice(int x1, int y1, int x2, int y2) {
		for (int i = y1; i<y2; i++) {
			for (int j = x1; j<x2; j++) {
				graph[i][j] = 1;
			}
		}
	}
	static void bfs(int x, int y) {
		Queue<int[]> q = new LinkedList<>();
		int area=0;
		
		
		q.add(new int[] {x,y});
		visited[x][y] = true; 
		area++;
		while (!q.isEmpty()) {
			int[] temp = q.poll();
			int tx = temp[0];
			int ty = temp[1];
			
			for(int i=0; i<4; i++) {
				int nx = tx + dx[i];
				int ny = ty + dy[i];
				
				if( 0<=nx && nx<m && 0<= ny && ny<n) {
					if(!visited[nx][ny] && graph[nx][ny] ==0) {
						visited[nx][ny] = true;
						area++;
						q.add(new int[] {nx,ny});
					}
				}
			}
		}
		result.add(area);
	}
}
